README.md
[English](/README.md) | [简体中文](/README.zh_CN.md)

<p align="center">
  <a href="" rel="noopener">
 <img width=200px height=200px src="img/PEPL-logo.png" alt="Project logo"></a>
</p>
<h2 align="center"><i>Part-of-speech Enhanced Prompt Learning</i>  for clickbait detection</h2>

<div align="center">

  [![License](https://img.shields.io/badge/license-MIT-blue.svg)](/LICENSE)
  ![Platform](https://img.shields.io/badge/platform-Windows%20%7C%20Linux-lightgrey)
  [![Python](https://img.shields.io/badge/python-3.10.8-brightgreen)](https://www.python.org/downloads/release/python-3108/)
  [![Pytorch](https://img.shields.io/badge/pytorch-1.13.0-orange)](https://pytorch.org/get-started/previous-versions/#v1130)
  ![Last Commit](https://img.shields.io/github/last-commit/WeSeewy/Chinese-Clickbait)


</div>

---



## 📝 Table of Contents
- [About](#about)
- [Getting Started](#getting_started)
- [Usage](#usage)
- [Authors](#authors)
- [Contributing](#contributing)

## 🧐 About <a name = "about"></a>
This project implements a detection method for Chinese clickbait (标题党) news, which can determine whether the  Chinese text belongs to the clickbait or not. Examples of clickbait and non-clickbait news are shown below:

|  Example   | Label  |
|  ----  | ----  |
| 当女生问“你在干啥”，这句回答100%勾起她的兴趣 <br/>When a girl asks "what are you doing", this answer is 100% to arouse her interest. | clickbait |
| 北京发布做好复工复产疫情防控常态化工作通告 <br/>Beijing government issued the Notice on Normalized Covid-19 Prevention and Control Measures for Resumption of Work.  | non-clickbait |

The main idea is in the research paper *Detecting Clickbait in Chinese Social Media by Prompt Learning*, which is accepted for publication in CSCWD'23.

## 🏁 Getting Started <a name = "getting_started"></a>
The project needs prepared work first before it can be used.

### Prerequisites
The dependency frameworks of this project are as follows:

- [Pytorch](https://pytorch.org/) - Deep Learning framework, please refer to the [guide](https://pytorch.org/get-started/previous-versions/#v1130) for installation, the version is 1.13.0
- [LTP](https://github.com/HIT-SCIR/ltp) - Chinese NLP tools, please refer to the [guide](https://github.com/HIT-SCIR/ltp) for installation, the version is 4.2.11
- [Transformers](https://github.com/huggingface/transformers) - Pretrained Language Model framework, please refer to the [guide](https://huggingface.co/docs/transformers/installation) for installation, the version is 4.24.0
- [OpenPrompt](https://github.com/thunlp/OpenPrompt) - Prompt Learning framework, please refer to the [guide](https://github.com/thunlp/OpenPrompt#installation) for installation, the version is 1.0.1

In addition, there are other dependency packages that need to be installed. If your python version is 3.10.8, you can install them directly by executing the following command:

```
pip install -r requirements.txt
```

If you are using another python version, please modify the requirements.txt to make it compatible.


### Dataset
For this project, the [WCD](https://github.com/natsusaikou/WeChat-Clickbait) dataset is for training and testing, please download the [all_labeled.csv](https://github.com/natsusaikou/WeChat-Clickbait/blob/master/data/all_labeled.csv) file and copy it in the **/data** path. If you must change the dataset store location, please modify the *DatasetPath* variable in the **main.py** file.


## 🎈 Usage <a name="usage"></a>
### Train

After completing the preparations, the following command can be used for training:

```
python main.py -s 0.01 0.5
```

The meaning of the two parameters for -s is using 1% of the dataset as the training set and 50% of the dataset as the test set.

The following command can be used to train models in extremely few-shot scenarios:

````
python main.py -f 16
````

The meaning of the parameter for -f is that only 16 clickbait and 16 non-clickbait samples are used for training.

In addition, the command supports other settings, as follows:

- -bs: set training batch size
- -lr: set training learning rate
- -ep: set training epoch
- -m: set mode，only support *pepl* and *base*，where *pepl* is our method，and *base* is the baseline
- -plm: set the PLM architecture，support *bert*、*roberta*、*ernie*、*Erlangshen*

An example of a completely set training command is as follows:

```
python main.py -s 0.01 0.5 -bs 8 -lr 5e-5 -ep 3 -m pepl -plm bert
```

### Predict

After the training, models will be stored in the **/checkpoints** folder. The input can be predicted by executing the following command, where the **items.txt** and **news.txt** files in the **/data** folder are samples to be predicted.

```
python use.py -p .\checkpoints\base_bert_fc_2.pt -m base -plm bert
```

The meanings of the parameters are shown below:

- -p: set models store location，such as the base_bert_fc_2.pt 
- -m: set mode，only support *pepl* and *base*
- -plm: set the PLM architecture，support *bert*、*roberta*、*ernie*、*Erlangshen*

**Note: -m and -plm should be the same as the settings used in training phase**


## ✍️ Authors <a name = "authors"></a>
- [@WeSeewy](https://github.com/WeSeewy) - Idea & Develop
- [@caomingpei](https://github.com/caomingpei) - Test & Doc

See also the list of [contributors](https://github.com/WeSeewy/Chinese-Clickbait/contributors) who participated in this project.

## ⛏️ Contributing <a name = "contributing"></a>
If you find problem(s) in this repo, please make issues.

If you want to contribute to this repo, please fork and create a new pull request.

The commit style should follow the convention: 

```
[!TYPE:] message
```

The [!TYPE:] includes the following types:

- !F: create new function
- !B: bug fix
- !D: update about the documents
- !S: change code style
- !R: refactor the code
- !O: optimize the performance
- !T: add test
- !C: chores of update the dependency
- !A: archive the related files

Example: 

```
!D: configuring the git style
```

This example means this commit [!D:] is about update a document and the commit reason is configuring the git style.


data/news.txt
{"id": "10013426", "title": "建国初的河南农村，拍电影《李双双》张瑞芳给了一个姑娘一件衣裳", "link": "http://www.sohu.com/a/321013876_120073393", "date": "Jun 17, 2019", "entity": ["郊区", "李双双", "张瑞芳", "新乡县", "新乡市", "新乡师范学院", "王屋山", "刘启云", "林县"], "content": "1961年著名电影《李双双》在林县拍摄，饰演李双双的著名演员张瑞芳在拍摄过程中和照片里的几位姑娘结成好友，右一姑娘身穿的衬衣就是张瑞芳所赠，而且正是张瑞芳戏中的装扮。1962年林县农村妇女闲不住，做衣做饭做针线。1957年 新乡市郊区丰收的喜悦。1958年 新乡县1956年 王屋山上农业方面全国劳模刘启云在给大学生讲课。1958年新乡师范学院上世纪50年代的结婚照。上世纪50年代干部结婚照 1958年 新乡市", "status": 0}
{"id": "10095086", "title": "普京成功拿捏住美国，拜登一改强硬态度，再三强调绝不能核战", "link": "http://www.sohu.com/a/513485241_600524", "date": "Dec 31, 2021", "entity": ["俄军", "美俄", "美国", "普京", "俄罗斯", "白宫", "拜登", "克林姆林宫", "苏联", "乌克兰"], "content": "普京成功拿捏住美国，拜登一改强硬态度，再三强调绝不能核战 文/空山在美俄紧张态势不断升级的背景下，普京和拜登日前进行了第2次视频会晤，据介绍，在会谈之初，拜登的态度十分强硬，扬言俄若敢进攻乌克兰，那么西方将“果断反应”，然而普京却没有拜登吓倒，反而更加强硬警告称，如果西方胆敢对俄实施所谓的“空前制裁”，那么美俄将彻底“撕破脸”，这直接让自诩为世界霸主的美国“破防”。事实证明，面对“战斗民族”的硬气，拜登的强硬都是装的，他迅速放低调门，再三向普京呼吁，美俄之间绝不能爆发核战争，强调核战不会有任何赢家。在此次“普拜会”结束后，克林姆林宫表示对结果“满意”，而白宫则顾左右而言其他，显然没有得偿所愿。虽然是全世界仅存的超级大国，然而美国也有最为忌惮的对手，就是俄罗斯。虽然从国力对比来看，俄只有美国的零头，但是却有着让美国寝食难安的强大核力量，这也是拜登这次“服软”的主要原因。根据最新的数据，俄军的核弹头总数约为1500枚，在规模上已经领先美国，而且普京多次强调，俄罗斯在必要时刻将首先使用核武器打击敌人，即便是蛮横惯了的美国，也不得不对俄方的警告三思。更让美国担心的是，俄军还继承了苏联时期的“死手系统”，这是旨在发动核报复的自动指挥控制系统。简单来说，就算美国对俄罗斯发动了先发制人的核打击，一口气向俄本土发射了数百枚核弹，摧毁了俄决策层和有生力量，但是“死手系统”将在无人下达命令的情况下，自动发射核导弹，对预定的目标进行疯狂的核报复。因此美国的讹诈可能对其他国家有效，但是对俄罗斯来说则完全没用，因为美国就算铤而走险，率先挑起核战争，也绝对无法在俄罗斯的核报复下幸存。在此次普拜会中，拜登一开始企图故伎重演，用伪装的强硬态度来吓唬普京，然而后者大风大浪可以说是见得太多了，直接就轻蔑拒绝，并且反过来威胁美国“别搞事”，直接让美国“破防”，拜登也不得不多次强调不想看到美俄爆发核冲突。事实证明，对付美国，还就得看“战斗民族”的这一套强硬风格。", "status": 0}
{"id": "10118778", "title": "通报：双双被免职！", "link": "http://www.sohu.com/a/513999245_349247", "date": "3 Jan 2022", "entity": ["崔诗越", "长安区委", "王斌", "雁塔区", "西安市", "雁塔区委", "西安", "崔诗", "病例", "国家卫健委", "杨建强", "未央区委", "副书记", "汉族", "雁塔区政府办公室"], "content": "据@央视新闻，为进一步加强西安市雁塔区疫情防控工作力量，经陕西省委同意，西安市副市长杨建强兼任雁塔区委书记，免去王斌同志雁塔区委书记职务，免去崔诗越同志雁塔区委副书记、区政府党组书记职务。杨建强，男，汉族，1968年7月生，今年54岁，研究生学历，工商管理硕士，中共党员。公开资料显示，杨建强历任西安市雁塔区曲江街道党（工）委副书记、办事处主任，雁塔区政府办公室主任，区委常委、副区长，长安区委副书记，副区长、代区长、区长，区委书记，未央区委书记、西安汉长安城国家大遗址保护特区党工委书记（兼）。2021年8月26日，在西安市十六届人大常委会第四十五次会议上，杨建强被任命为西安市副市长。据西安市政府官网介绍，杨建强负责科技、工业、生态环境、国有资产监管、大数据、金融等方面的工作。分管市科学技术局（市外国专家局）、市工业和信息化局、市生态环境局、市国有资产监督管理委员会、市大数据资源管理局、市金融工作局。在此次被免职之前，雁塔区委书记王斌、区长崔诗越上任均一年左右。王斌，男，汉族，1973年10月生，甘肃平凉人，1995年6月加入中国共产党，1996年7月参加工作，全日制大学学历，法学学士。公开资料显示，王斌长期在西安工作，担任过西安航天基地党工委副书记、管委会副主任，雁塔区区长等职。2021年1月7日，王斌任中共西安市雁塔区委书记。王斌履新的同一天（2021年1月7日），崔诗越任西安市雁塔区委委员、常委、副书记。崔诗越，男，汉族，1974年10月生，研究生学历，工学博士，中共党员。公开资料显示，崔诗越担任过市委办公厅秘书三处副处级秘书，市纪委办公厅副主任，未央区委常委、统战部部长，未央区委常委、区纪委书记、区监察委员会主任，碑林区委副书记，雁塔区委副书记、区人民政府党组书记、副区长、代区长，市小寨地区综合改造管委会党组书记、主任（兼）。2021年3月，他当选为西安市雁塔区区长。王斌、崔诗越是雁塔区疫情防控指挥部总指挥长。他们两人的最后一次公开亮相，是在去年12月26日。新闻延伸据国家卫健委1月2日最新通报，1月1日0—24时，陕西新增本土病例123例，其中西安市122例、延安市1例。1月2日下午，陕西西安市召开新冠肺炎疫情防控工作新闻发布会。会上通报，2022年1月1日0时—24时，西安新增本土确诊病例122例。2021年12月9日0时—2022年1月1日24时，西安累计报告本土确诊病例1573例。目前，所有病例均已转运到定点医院隔离治疗。新增122例确诊病例中，均为轻型；从发现方式看，隔离管控中发现104例、核酸筛查中发现18例；从居住地分布看，雁塔区61例、高新区23例、西咸新区16例、长安区9例、碑林区7例、莲湖区4例、国际港务区1例、鄠邑区1例；从年龄结构看，3岁以下4例、3—18岁15例、18—60岁94例、60岁以上9例；从性别分布看，男性66例、女性56例。根据流调，已排查出的密接和次密接均按照有关要求落实隔离管控。西安社区传播风险依然存在聚集是当前西安疫情的最大风险发布会上，西安市疾控中心主任陈保忠介绍，目前从近几轮核酸筛查数据来看，从社区检出的病例还有一定比例，社区传播风险依然存在。西安市本土疫情正处于胶着对垒的关键阶段。聚集是当前疫情的最大风险。西安全市在核酸检测时，全面推行分人群、分时段、分区域、预约采样，坚决杜绝居家隔离等风险人群与其他人群发生沾染的现象。在购买、领取生活物资等物品时，建议戴上手套，一定不要拥挤、扎堆。高层住宅的住户，下楼时帮助老人小孩戴好口罩做好防护，在电梯内避免说话、打闹，不用手直接接触电梯按钮等部位，回家后先洗手，可对外穿衣物袖口等部位喷洒75%的酒精，再摘除口罩，口罩可挂在卫生间窗口附近，最好每次参加核酸检测后更换口罩。来源 | 央视新闻、国家卫健委官网、陕西日报、新华社、西安晚报", "status": 0}
{"id": "10096771", "title": "原创\n2021最后一天，法国日增破欧洲记录，美日增破70万", "link": "http://www.sohu.com/a/513529838_120504280", "date": "03-Jan-2022", "entity": ["马萨诸塞州", "约翰逊政府", "赵立坚", "美国", "科林·鲍威尔", "莫德纳", "法国", "莫纳德", "印度", "联邦政府", "拜登政府", "拜登", "美国疾病控制与预防中心", "欧洲", "英国"], "content": "欧洲正迎来“奥密克戎海啸”，继英国之后，法国刷新欧洲最高纪录，从日增超10万到超20万，只用了4天时间。法国防疫措施再次收紧，巴黎已开始实施户外口罩强制令，民众必须按规定天数居家办公，违规要罚款。专家预计马上到来的1月份，法国日增将飙升至25万。20万+！以为这个在我们看来似乎不可理喻的数字是极限了吗？并不是！还有美国。也是仅仅4天时间，美国日增从超50万到了超70万。如此令人听之色变的数字，在美国国内并未掀起任何波澜。拜登一边依然淡定地喊大家一定要打疫苗，另一边忙着和各个州相互推卸责任，疫情的锅就这样被甩来甩去。实际也不难理解，疫情以来，美国就没正常过，没尝到“清零”后的甜头，自然也不会明白正常生活的意义，况且根本还没能力“清零”。都已经日增70万+了，美国国家防疫指南居然还把隔离时间从10天缩减到了5天，令人叹为观止的是，仍有近40%的美国人属于拒绝接种疫苗的顽固分子。美国媒体做了一个统计，本周美国已报告了超过190万例新增确诊病例，这是什么概念？就是相当于每秒钟有约3名美国人感染。美国累计死亡病例已超84万。美国疾病控制与预防中心预测，在未来一个月内，美国或将有超过4.4万人死于新冠肺炎。对于全球死亡人数最高这顶“桂冠”，美国媒体表示不同意，他们注意到印度公布的实际死亡人数和官方公布的有巨大出入。美媒还做了粗略统计，印度实际死亡人数很可能已超600万，这样美国超80万的数据很明显就不够看了！美国媒体还有心情为印度操心，平时争第一跑得比谁都快，这时候反倒是谦虚起来了。超80万的死亡率可当真应了“人命如草芥”这句话，如果印度的统计数据基本上等于“不统计”，那么美国的防疫政策也就是“不防疫”。并且，美国媒体真有自信觉得自己国家的检测数据是真实的吗？美国落后的检测能力甚至还不如英国，约翰逊政府七上八下的防疫政策数次被人诟病，如今英国每天每千人只能进行22.3次新冠病毒检测，美国的检测能力是英国的五分之一。拜登政府还拒绝了一项旨在每月向美国民众提供7.32亿次核酸检测能力的提议。别问为什么，问就是疫情不能只靠联邦政府，各个州政府动起来解决事情。美国人肯定也已经习惯了，各个州各自为政一盘散沙，民众种族分裂政见不合，隔三差五爆出来一个枪击案，死亡已经成了家常便饭。民众不待见政府，不相信政府，自然也不会在困难时期想到政府，不听劝导也就是自然而然的结论。看看那些死去的人，有几个社会名流、政客、企业家这些精英在疫情高发期死亡的？死去的大多数都是老百姓，这些老百姓中大部分还是中老年人。没钱没地位年纪也大了，这些名词串在一起就成了美国社会的底层，只有美国前国务卿科林·鲍威尔这位为全球所熟知的“洗衣粉爱好者”，10月因新冠并发症去世时引起了一阵水花，但84岁的鲍威尔终究只是以为过气政客。同一时期，美国疫苗公司莫德纳的几位高层首次登上了福布斯富豪榜。一面是疫苗分配不均，一面是美国疫情浪费严重莫纳德靠着疫苗狠赚了200亿美元，朱门酒肉臭，谁会管路有冻死骨。美国疫情不容乐观，制药企业开足马力，美国有些州自然也找上了中国企业。有中企表示收到了美国马萨诸塞州等地政府超14亿人民币的订单。不过如今中企在与美国打交道时也非常小心谨慎，鉴于美国有些企业中途取消订单及拖欠付款的黑历史，如今许多中企多长了心眼，付款再上马，付多少钱出多少货，尾款不能欠太多。看到美国疫情如此严重，医疗系统不堪重负，死亡数据冠绝全球，通胀严重，国内还在不停地印钱，美国快不行了？美元快崩了？美国还有救吗？实际上，美国的全球霸权不是一朝一夕搭建起来的，自然也不可能因为一场疫情就塌了。就像嚷嚷着减肥的人，身上的肉是自己一天一天吃起来的，也不可能在短时间内都甩掉，必将付出巨大的艰辛。有意思的是，12月30日，我国外交部发言人在回应记者“有拉美人士向世卫组织写信谴责其关于疫情方面的不力反应”时，赵立坚表示，中方在新冠肺炎疫情问题上无可指责。“在座的外国记者，能在抗击疫情期间生活在中国，你们就偷着乐吧。”相信感同身受的不会只有赵立坚。万丈高楼平地起，美国富人政客们还在狂欢，看不见的是根基已经摇摇欲坠了！", "status": 0}

main.py
import argparse
import os

import bert
import prompt
from helper import utils
from preprocess import ltp_tag_dataset
import torch
import warnings
warnings.filterwarnings("ignore")

TrainScale = 0.02       # The scale of train dataset
DevScale = 0.5          # The scale of test dataset
LearningRate = 5e-5     # The learning rate for AdamW
Epoch = 3               # Training epoch
RandomSeed = 42         # Random seed

DatasetPath = os.path.join("data", "all_labeled.csv")
LTPDatasetPath = os.path.join("data", "ltp_labeled.csv")

if __name__ == '__main__':
    arg_description = "This script is an implement of PEPL(Part-of-speech Enhanced Prompt Learning) method for " \
                      "clickbait news detection. Our work has been accepted by [CSCWD'23]."
    parser = argparse.ArgumentParser(description=arg_description)

    group = parser.add_mutually_exclusive_group()
    group.add_argument('--scale', '-s', nargs=2, type=float, help='Need 2 arguments to set the train and test scale of '
                                                                  'the dataset. The first input is for train and the '
                                                                  'second input is for test. For example, -s 0.01 0.5 '
                                                                  'means 1% of the dataset used for training, 50% of '
                                                                  'the dataset used for testing and others used '
                                                                  'for validation.')
    group.add_argument('--few_shot', '-f', type=int, help='Used for set few shot scenario. For example -f 16 means that'
                                                          'only 16 items in the dataset are used for training.')

    parser.add_argument('--batch_size', '-bs', type=int, default=8, help='the batch size used in training')
    parser.add_argument('--learning_rate', '-lr', type=float, default=LearningRate, help='learning rate for AdamW')
    parser.add_argument('--training_epoch', '-ep', type=int, default=Epoch, help='training epoch setting')
    parser.add_argument('--mode', '-m', type=str, default='pepl', choices=['pepl', 'base'],
                        help='the mode used for method, the pepl is default and the base is for baseline.')
    parser.add_argument('--pretrained_language_model', '-plm', type=str, default='bert',
                        choices=['bert', 'roberta', 'ernie', 'Erlangshen'],
                        help='the pretrained language model used for training and testing.')

    args = parser.parse_args()

    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
    if args.mode == 'pepl':
        if not os.path.exists(LTPDatasetPath):
            DatasetPath = ltp_tag_dataset(DatasetPath)
        dataset = utils.wechat_only2_dataset(DatasetPath)
        prompt.run(device, RandomSeed, dataset, args.scale, args.few_shot, args.batch_size, args.learning_rate,
                   args.training_epoch, args.mode, args.pretrained_language_model)
    elif args.mode == 'base':
        dataset = utils.wechat_only2_dataset(DatasetPath)
        bert.run(device, RandomSeed, dataset, args.scale, args.few_shot, args.batch_size, args.learning_rate,
                 args.training_epoch, args.mode, args.pretrained_language_model)
    else:
        raise RuntimeError("Invalid mode, please check")